src/settings/SettingsStore.ts

```diff
@@ -719,24 +719,41 @@ export default class SettingsStore {
     /**
      * Migrate the setting for visible images to a setting.
      */
-    private static migrateMediaControlsToSetting(): void {
-        const MIGRATION_DONE_FLAG = "mx_migrate_media_controls";
-        if (localStorage.getItem(MIGRATION_DONE_FLAG)) return;
+    private static migrateMediaControlsToSetting(isFreshLogin: boolean): void {
+        if (isFreshLogin) return;
+        const client = MatrixClientPeg.safeGet();
 
-        logger.info("Performing one-time settings migration of show images and invite avatars to account data");
-        const handler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
-        const showImages = handler.getValue("showImages", null);
-        const showAvatarsOnInvites = handler.getValue("showAvatarsOnInvites", null);
+        const doMigration = async (): Promise<void> => {
+            // Never migrate if the config already exists.
+            const existingConfig = client.getAccountData("io.element.msc4278.media_preview_config");
 
-        const AccountHandler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
-        if (showImages !== null || showAvatarsOnInvites !== null) {
-            AccountHandler.setValue("mediaPreviewConfig", null, {
-                invite_avatars: showAvatarsOnInvites === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
-                media_previews: showImages === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
-            });
-        } // else, we don't set anything and use the server value
+            if (existingConfig) {
+                return;
+            }
+            logger.info("Performing one-time settings migration of show images and invite avatars to account data");
+            const handler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
+            const showImages = handler.getValue("showImages", null);
+            const showAvatarsOnInvites = handler.getValue("showAvatarsOnInvites", null);
+
+            if (typeof showImages === "boolean" || typeof showAvatarsOnInvites === "boolean") {
+                this.setValue("mediaPreviewConfig", null, SettingLevel.ACCOUNT, {
+                    invite_avatars: showAvatarsOnInvites === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
+                    media_previews: showImages === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
+                });
+            } // else, we don't set anything and use the server value
+        };
 
-        localStorage.setItem(MIGRATION_DONE_FLAG, "true");
+        const onSync = (state: SyncState): void => {
+            if (state === SyncState.Prepared) {
+                client.removeListener(ClientEvent.Sync, onSync);
+
+                doMigration().catch((e) => {
+                    logger.error("Failed to migrate URL previews in E2EE rooms:", e);
+                });
+            }
+        };
+
+        client.on(ClientEvent.Sync, onSync);
```

Comment:
a sync may have already happened by this point, as `runMigrations` happens after `await startMatrixClient` - I suggest making use of `if (!client.isInitialSyncComplete()) ...` to gate the sync listener

```suggestion
		while (!client.isInitialSyncComplete()) {
			await emitPromise(client, "sync");
		}
		
		doMigration().catch((e) => {
            logger.error("Failed to migrate URL previews in E2EE rooms:", e);
        });
```

or similar

Reply:
Done.